python/cudf/cudf/io/parquet.py

```diff
@@ -534,6 +536,146 @@ def write_to_dataset(
+      - Each clause's predicates into a conjunction (AND)
+      - All clauses into a disjunction (OR) of those conjunctions
+    """
+
+    def _raise_for_unsupported_scalar_types(val: Any):
+        if not (
+            isinstance(val, (int, float))
+            or isinstance(val, (datetime.date, datetime.timedelta))
+            or isinstance(val, str)
+        ):
+            return val
+        raise NotImplementedError(
+            "Only numeric, string, or timestamp/duration scalars are accepted"
+        )
+
+    def make_expr(col: str, op: str, val: Any) -> plc_expr.Expression:
+        col_ref = plc_expr.ColumnNameReference(col)
+
+        if op in _BIN_OPS:
+            return plc_expr.Operation(
+                _BIN_OPS[op],
+                col_ref,
+                plc_expr.Literal(
+                    plc.Scalar.from_py(
+                        _raise_for_unsupported_scalar_types(val)
+                    )
+                ),
+            )
+        elif op == "in":
+            values = [
+                plc_expr.Literal(
+                    plc.Scalar.from_py(_raise_for_unsupported_scalar_types(v))
+                )
+                for v in val
+            ]
+            return reduce(
+                lambda acc, x: plc_expr.Operation(
+                    plc_expr.ASTOperator.LOGICAL_OR, acc, x
+                ),
+                [
+                    plc_expr.Operation(plc_expr.ASTOperator.EQUAL, col_ref, v)
+                    for v in values
+                ],
+            )
+        elif op == "not in":
+            values = [
+                plc_expr.Literal(
+                    plc.Scalar.from_py(_raise_for_unsupported_scalar_types(v))
+                )
+                for v in val
+            ]
+            return reduce(
+                lambda acc, x: plc_expr.Operation(
+                    plc_expr.ASTOperator.LOGICAL_AND, acc, x
+                ),
+                [
+                    plc_expr.Operation(
+                        plc_expr.ASTOperator.NOT_EQUAL, col_ref, v
+                    )
+                    for v in values
+                ],
+            )
+        elif op == "is":
+            if val is None:
+                return plc_expr.Operation(
+                    plc_expr.ASTOperator.IS_NULL, col_ref
+                )
+            raise NotImplementedError("Only `is None` supported")
+        elif op == "is not":
+            if val is None:
+                return plc_expr.Operation(
+                    plc_expr.ASTOperator.NOT,
+                    plc_expr.Operation(plc_expr.ASTOperator.IS_NULL, col_ref),
+                )
+            raise NotImplementedError("Only `is not None` supported")
+        else:
+            raise NotImplementedError(f"Unsupported op: {op}")
+
+    clauses = [
+        reduce(
+            lambda acc, x: plc_expr.Operation(
+                plc_expr.ASTOperator.LOGICAL_AND, acc, x
+            ),
```

Comment:
*suggestion*: this is more neatly written as:
```
from functools import partial
...

clauses = [reduce(partial(plc_expr.Operation, plc_expr.ASTOperator.LOGICAL_AND),
                  (make_expr(*pred) for pred in clause)) for clause in filters]
```
I think.

Reply:
Done